### PR TITLE
New Full Logbook

### PR DESCRIPTION
**Pull request by PTschaikner**
### Summary

I have created a feature to extract a more exhaustive logbook for Aurora boards. You can now retrieve a CSV file that includes attempts on boulders that haven't been climbed yet.

### New Logbook Fields

- **tries:** Returns the tries in the session.
- **tries_total:** Returns all tries on the same boulder (also taking angle and is_mirror into account).
- **sessions_count:** Returns the number of sessions on the boulder.
- **logged_grade:** Returns the grade that has been logged by the user.
- **displayed_grade:** Returns the grade provided by the app (integrated because the difficulty is only logged when a climb has been completed successfully).
- **is_ascent:** Indicates whether the boulder was climbed in this session.

These features represent the information I’d like to use for my own analysis, but I believe they might be useful for others as well.
### Problems/Issues/Questions
- **displayed_grade Limitation:** The displayed_grade only works when a local database is provided. I haven't found an elegant way to retrieve this information from the API without significant performance issues. One potential solution is to integrate results from get_climb_stats and use the most logged or median difficulty.
- **Testing Scope:** The feature has only been tested with my own data from Grasshopper and Tension boards. More extensive testing might be needed to ensure robustness.
- **Backward Compatibility:** There might be issues with how tries were saved in older versions where the mechanics for "tries" were different. My database doesn't include such sends, making it difficult to test scenarios where the lookup returns something like "3 months".
- **Code Quality:** The code is not very elegant at this point and could benefit from further refactoring and cleanup.

### Recommendations

To balance functionality and performance, I propose maintaining two separate functions:

- Basic Logbook: For general use without a requirement for a local database.

- Detailed Logbook: For more in-depth analysis, requiring a local database to avoid performance issues.

Currently, the full logbook returns NA for displayed_grade when no local database is available, resulting in slow performance. Using a local database improves speed significantly.
### Conclusion

If these features align with your goals for the project, I am happy to refine the code based on your feedback and suggestions. Thank you for the opportunity to contribute. I look forward to building visualizations with this data.

### CODE REVIEW

File: src/boardlib/api/aurora.py
Comment (lemeryfertitta) [inline src/boardlib/api/aurora.py:479]: When is "Unknown Climb" occurring? If this isn't something that we expect to happen normally, I'd prefer to just leave it as is. It seems like we should raise errors if we can't find the climb name.

File: src/boardlib/api/aurora.py
Comment (lemeryfertitta) [inline src/boardlib/api/aurora.py:549]: Prefer `None`

File: src/boardlib/api/aurora.py
Comment (lemeryfertitta) [inline src/boardlib/api/aurora.py:556]: Prefer `None`

File: src/boardlib/api/aurora.py
Comment (lemeryfertitta) [inline src/boardlib/api/aurora.py:442]: This function feels a bit too large. Could we break it into a few logical chunks using helper functions?

**Review by lemeryfertitta — CHANGES_REQUESTED** [review]
Thank you for the PR!

To address some of your concerns/questions:

> The displayed_grade only works when a local database is provided

That seems fine to me for now, let's just make a note of it in the command help and README.

> Testing Scope: The feature has only been tested with my own data from Grasshopper and Tension boards. More extensive testing might be needed to ensure robustness.

I will test it out at some point and make sure it plays nice with the old ascents

> Code Quality: The code is not very elegant at this point and could benefit from further refactoring and cleanup.

Made a few suggestions, we can iterate on this a bit, I'm happy to help.

> I propose maintaining two separate functions:

I think that I prefer one function that does it all, but with the optional database as you've provided already.

> I look forward to building visualizations with this data.

If you end up building any cool visualizations that seem like they may fit within the library, feel free to push those here! I think others may find them useful.

File: src/boardlib/api/aurora.py
Comment (PTschaikner) [reply src/boardlib/api/aurora.py]: should never happen, changed.

File: src/boardlib/api/aurora.py
Comment (PTschaikner) [reply src/boardlib/api/aurora.py]: Changed to None, tried to add a better help command.

File: src/boardlib/api/aurora.py
Comment (PTschaikner) [reply src/boardlib/api/aurora.py]: I changed the value to None and added logic to include V-scale grade names. If you log an unclimbed project, it doesn't get a difficulty value here, so using .get seems appropriate here

File: src/boardlib/api/aurora.py
Comment (PTschaikner) [reply src/boardlib/api/aurora.py]: .get was not appropriate here, changed.

File: src/boardlib/api/aurora.py
Comment (PTschaikner) [reply src/boardlib/api/aurora.py]: agreed and changed.

File: src/boardlib/api/aurora.py
Comment (PTschaikner) [reply src/boardlib/api/aurora.py]: done

File: src/boardlib/api/aurora.py
Comment (PTschaikner) [reply src/boardlib/api/aurora.py]: I tried to refactor the function, I think it's a bit more clear now.

**Comment by lemeryfertitta** [comment]
Tested locally - I actually have two accounts, one from pre-bids and one from post-bids. I'm not sure if this is something that all pre-bids users had to do or if it was just me forgetting my account.

Anyways, post-bids behavior is working well. Pre-bids fails with the following error:

```
File ".../boardlib/api/aurora.py", line 399, in get_bids_logbook
    return sync_results["PUT"]["bids"]
KeyError: 'bids'
```

So maybe we can just try/except on the attempt to get bids to fall back on the current behavior in that case. Once that error is fixed and https://github.com/lemeryfertitta/BoardLib/pull/38#discussion_r1636935396 is addressed, I think this PR will be good to go and we can get these features released.

**Comment by PTschaikner** [comment]
Thanks for testing! I was able to recreate your error by providing a database with an empty bids table. There's a similar issue with an empty ascents table, though it's unlikely to occur. In both cases, I think returning an empty table would be better than raising an unspecific error. If you have the time, it would be great if you could test whether this fix works with your old account.

Before merging the logbook commands, it would be helpful if you could test the basic functionality of the moon logbook. I created an account with a dummy logbook using the current moonboard app, but I couldn't get it to work with your original command, even after rolling back to an older commit.

Are the tests up to date? I encountered several errors when running them. I'd like to experiment with writing my first tests, and this project seems like a good starting point, if you don't mind.

**Comment by lemeryfertitta** [comment]
> In both cases, I think returning an empty table would be better than raising an unspecific error. If you have the time, it would be great if you could test whether this fix works with your old account.

Good idea, better than my suggestion. I tested your fix and it works perfectly for both of my accounts.

> Before merging the logbook commands, it would be helpful if you could test the basic functionality of the moon logbook. I created an account with a dummy logbook using the current moonboard app, but I couldn't get it to work with your original command, even after rolling back to an older commit.

I tested and the moon logbook is indeed broken. A quick investigation suggests that moon has recently changed their site. The API looks to be mostly the same but I think they changed the login flow. I'll create a separate ticket for this. Since that is broken anyways, let's still merge your functionality into the main `logbook` command. Don't worry about breaking Moon more since it is already broken.

> Are the tests up to date? I encountered several errors when running them. I'd like to experiment with writing my first tests, and this project seems like a good starting point, if you don't mind.

Please feel free to improve/add to the tests. I used them when I worked on the first release of this library but as changes were made they were ignored. If you are motivated, it would be great to get them fixed up and eventually add them to Github CI and potentially block merging for changes with failing tests.

**Comment by PTschaikner** [comment]
I have made the following updates in this Pull Request:

- **Unified Logbook Command:** Integrated the new version for a more streamlined experience.
- **Bug Fixes:** Resolved minor bugs, including issues with grade conversion.
- **Updated Unit Tests:** Ensured the unit tests are up to date and functional.
- **Documentation:** Added detailed information about the new logbook features in the `README.md`.

Future Improvements I Plan to Work On:

- **Enhanced Error Handling:** Focusing on login errors and issues with incorrect database files.
- **Expanded Unit Tests:** Planning to add more comprehensive tests to improve reliability.
- **Visualizations:** Developing D3.js visualizations for logbook data. I aim to create a solution that allows users to view a static .html file without needing to run a local server.

**Review by lemeryfertitta — APPROVED** [review]
Great, looking good to me! I'm happy to get this one merged and released, or we can keep it open for any of the future improvements you had in mind. Let me know what you prefer.

**Comment by PTschaikner** [comment]
Feel free to merge, especially if it is useful for Climbdex as suggested by @gardaholm. I hope it doesn't cause any headaches there. I'll focus on the visualizations and open a separate PR if I manage to get to a point that seems useful in this regard.

File: src/boardlib/api/aurora.py
Comment (gardaholm) [inline src/boardlib/api/aurora.py:547]: can we go straight for the token here instead of username, password? For climbdex we only save the bearer token after the initial login

File: src/boardlib/api/aurora.py
Comment (gardaholm) [inline src/boardlib/api/aurora.py:572]: adding the climb_uuid would be pretty helpful here
This would simplify the attempts lookup for climbdex a lot: https://github.com/lemeryfertitta/Climbdex/issues/55

`full_logbook_df = pd.DataFrame(final_logbook, columns=['climb_uuid', 'board', 'angle', 'climb_name', 'date', 'logged_grade', 'displayed_grade', 'difficulty', 'tries', 'is_mirror', 'is_ascent', 'comment'])`

File: src/boardlib/api/aurora.py
Comment (gardaholm) [inline src/boardlib/api/aurora.py:576]: `full_logbook_df = full_logbook_df.groupby(['climb_uuid', 'climb_name', 'is_mirror', 'angle']).apply(calculate_tries_total).reset_index(drop=True)`

File: src/boardlib/api/aurora.py
Comment (gardaholm) [inline src/boardlib/api/aurora.py:578]: `full_logbook_df['is_repeat'] = full_logbook_df.duplicated(subset=['climb_uuid', 'climb_name', 'is_mirror', 'angle'], keep='first')`

File: src/boardlib/api/aurora.py
Comment (lemeryfertitta) [reply src/boardlib/api/aurora.py:572]: I think that is a fine suggestion to make this more useful for Climbdex, but I would want to make sure we don't do that in a way that clutters the output of the boardlib `logbook` command. Realistically, users of the command line version don't need the climb UUID in the output CSV.

One idea would be to refactor out a version of this function that does everything you've suggested here, and then in the command-line version we can handle the extra steps. From my understanding, that would just be to login and get a token and also to filter out the climb_uuid. I'd be happy to accept a PR for this if it would help with your proposed climbdex additions!

Comment (gardaholm) [reply src/boardlib/api/aurora.py:572]: Alright, will do it this way!

**Pull request merged**
